File: ios/Vibree/Vibree/Utils/BertTokenizer.swift
import Foundation

class BertTokenizer {
    private var vocab: [String: Int] = [:]
    
    init() {
        if let path = Bundle.main.path(forResource: "vocab", ofType: "txt") {
            do {
                let content = try String(contentsOfFile: path)
                let lines = content.components(separatedBy: .newlines)
                for (index, line) in lines.enumerated() {
                    let word = line.trimmingCharacters(in: .whitespaces)
                    if !word.isEmpty {
                        vocab[word] = index
                    }
                }
            } catch {
                print("Error loading vocab: \(error)")
            }
        }
    }
    
    func tokenize(_ text: String, maxLength: Int = 128) -> (inputIds: [Int64], attentionMask: [Int64]) {
        var tokens: [Int64] = []
        var inputIds = [Int64](repeating: 0, count: maxLength)
        var attentionMask = [Int64](repeating: 0, count: maxLength)
        
        tokens.append(Int64(vocab["[CLS]"] ?? 101))
        
        let words = text.lowercased().components(separatedBy: CharacterSet.whitespacesAndNewlines)
        for word in words {
            if tokens.count >= maxLength - 1 { break }
            if let id = vocab[word] {
                tokens.append(Int64(id))
            } else {
                tokens.append(Int64(vocab["[UNK]"] ?? 100))
            }
        }
        
        if tokens.count < maxLength {
            tokens.append(Int64(vocab["[SEP]"] ?? 102))
        }
        
        for i in 0..<tokens.count {
            if i < maxLength {
                inputIds[i] = tokens[i]
                attentionMask[i] = 1
            }
        }
        
        return (inputIds, attentionMask)
    }
}

File: android/app/src/main/java/com/example/wearablecollector/logic/BertTokenizer.kt
package com.example.wearablecollector.logic

import android.content.Context
import java.io.BufferedReader
import java.io.InputStreamReader

class BertTokenizer(context: Context) {
    private val vocab = mutableMapOf<String, Int>()
    
    init {
        try {
            val inputStream = context.assets.open("vocab.txt")
            val reader = BufferedReader(InputStreamReader(inputStream))
            var index = 0
            reader.forEachLine { line ->
                vocab[line.trim()] = index
                index++
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    fun tokenize(text: String, maxLength: Int = 128): Pair<LongArray, LongArray> {
        val tokens = mutableListOf<Long>()
        val inputIds = LongArray(maxLength) { 0 }
        val attentionMask = LongArray(maxLength) { 0 }
        
        // [CLS] token
        tokens.add(vocab["[CLS]"]?.toLong() ?: 101L)
        
        val words = text.lowercase().split(Regex("\\s+"))
        for (word in words) {
            if (tokens.size >= maxLength - 1) break
            
            // Simplified WordPiece
            if (vocab.containsKey(word)) {
                tokens.add(vocab[word]!!.toLong())
            } else {
                // TODO: Full WordPiece subword splitting if needed
                tokens.add(vocab["[UNK]"]?.toLong() ?: 100L)
            }
        }
        
        // [SEP] token
        if (tokens.size < maxLength) {
            tokens.add(vocab["[SEP]"]?.toLong() ?: 102L)
        }
        
        for (i in tokens.indices) {
            if (i < maxLength) {
                inputIds[i] = tokens[i]
                attentionMask[i] = 1L
            }
        }
        
        return Pair(inputIds, attentionMask)
    }
}
